documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    /// <param name="position">The vector to transform.</param>
+    /// <param name="matrix">The transformation matrix.</param>
+    /// <returns>The transformed vector.</returns>
+    public static Vector4D<T> Transform<T>(Vector2D<T> position, Matrix4X4<T> matrix) where T : INumberBase<T>;
+
+    /// <summary>Transforms a two-dimensional vector by the specified Quaternion<T> rotation value.</summary>
+    /// <param name="value">The vector to rotate.</param>
+    /// <param name="rotation">The rotation to apply.</param>
+    /// <returns>The transformed vector.</returns>
+    public static Vector4D<T> Transform<T>(Vector2D<T> value, Quaternion<T> rotation) where T : ITrigonometricFunctions<T>, IRootFunctions<T>;
+
+    /// <summary>Transforms a three-dimensional vector by a specified 4x4 matrix.</summary>
+    /// <param name="position">The vector to transform.</param>
+    /// <param name="matrix">The transformation matrix.</param>
+    /// <returns>The transformed vector.</returns>
+    public static Vector4D<T> Transform<T>(Vector3D<T> position, Matrix4X4<T> matrix) where T : INumberBase<T>;
+
+    /// <summary>Transforms a three-dimensional vector by the specified Quaternion{T} rotation value.</summary>
+    /// <param name="value">The vector to rotate.</param>
+    /// <param name="rotation">The rotation to apply.</param>
+    /// <returns>The transformed vector.</returns>
+    public static Vector4D<T> Transform<T>(Vector3D<T> value, Quaternion<T> rotation) where T : ITrigonometricFunctions<T>, IRootFunctions<T>;
+
+    /// <summary>Transforms a four-dimensional vector by a specified 4x4 matrix.</summary>
+    /// <param name="vector">The vector to transform.</param>
+    /// <param name="matrix">The transformation matrix.</param>
+    /// <returns>The transformed vector.</returns>
+    public static Vector4D<T> Transform<T>(Vector4D<T> vector, Matrix4X4<T> matrix) where T : INumberBase<T>;
+}
+```
+
+IVector base type:
+```cs
+/*
+ * Explanation for this interface:
+ * If IDivisionOperators<TVector, T, TVector>/IMultiplyOperators<TVector, T, TVector> are implemented via IVector, they conflict with
+ * IDivisionOperators<TVector, TVector, TVector> from INumberBase<TVector> IMultiplyOperators<TVector, TVector, TVector>, which we need.
+ * This causes the following compiler error:
+ *
+ * Interface 'GenericVector.IVector<TVector,T>' cannot implement both 'System.Numerics.IDivisionOperators<TVector,TVector,TVector>' and
+ * 'System.Numerics.IDivisionOperators<TVector,T,TVector>' because they may unify for some type parameter substitutions
+ *
+ * As we can't add a "not-constraint" to generics, we have to live with this. An alternative is to not implement the interface but still
+ * provide the operator; this could be discussed with the working group.
+ */
+public interface IVectorAlso<TVector, T> :
+    IDivisionOperators<TVector, T, TVector>,
+    IMultiplyOperators<TVector, T, TVector>
```

Comment:
For the BCL types, we plan on simply not having a definition for `IDivisionOperators<TVector, T, TVector>`

Rather, we'd just define the operator itself directly on `IVector` and they definitions will be folded together if the types happen to be the same.

This shouldn't be problematic in practice, especially in generic contexts where the conversions between T and U need to be more explicit.

Reply:
I was mixed on whether to use the interfaces (to match convention) or just define the operators on IVector (to make the code less disturbing). Still undecided on that one.

Reply:
You risk some diamond problems when two interfaces define what can potentially resolve down to the same method when types line up.

this is a non issue when it’s a derived interface defining an overload that could resolve to the same as a method on a base interface, since they then become the same definition